File: solutions/Leetcode_91/leetcode_91.py
class Solution1:
    def num_decodings(self, s: str) -> int:
        valid_nums = {str(x): 1 for x in range(1, 27)}
        pprev = 1
        prev = 1 if s[0] in valid_nums else 0
        for i in range(1, len(s)):
            current = valid_nums.get(s[i], 0) * prev + valid_nums.get(s[i-1:i+1], 0) * pprev
            pprev, prev = prev, current
        return prev


class Solution2:
    def num_decodings(self, s):
        pprev = 1
        prev = 0 if s[0] == '0' else 1
        for i in range(1, len(s)):
            current = prev * (s[i] != '0') + pprev * (s[i-1] == '1' or s[i-1] == '2' and s[i] < '7')
            pprev, prev = prev, current
        return prev

File: solutions/Leetcode_91/leetcode_91.cpp
#include <string>
#include <unordered_map>

class Solution1
{
public:
    int numDecodings(std::string s)
    {
        std::unordered_map<std::string, int> validNums;
        for (int x = 1; x <= 26; x++)
            validNums[std::to_string(x)] = 1;
        int pprev = 1;
        int prev = validNums[s.substr(0, 1)] ? 1 : 0;
        for (int i = 1; i < s.size(); i++)
        {
            int current = validNums[s.substr(i, 1)] * prev + validNums[s.substr(i - 1, 2)] * pprev;
            pprev = prev;
            prev = current;
        }
        return prev;
    }
};

class Solution2
{
public:
    int numDecodings(std::string s)
    {
        int pprev = 1;
        int prev = s[0] == '0' ? 0 : 1;
        for (int i = 1; i < s.size(); i++)
        {
            int current = (s[i] != '0' ? prev : 0) + ((s[i - 1] == '1' || s[i - 1] == '2' && s[i] < '7') ? pprev : 0);
            pprev = prev;
            prev = current;
        }
        return prev;
    }
};
